feat: Sum and average aggregation queries

danieljbruce opened the pull request:
This PR demonstrates how we are going to leverage the current framework to support sum/average queries for later use.

danieljbruce committed:
Initial sum aggregation

This commit introduces the sum aggregation with a simple test to ensure it works

danieljbruce committed:
Modify encoding

This change modifies the encoding so that the right data reaches the grpc layer

danieljbruce committed:
PropertyAggregateField with tests

Adds helper functions and a super class so that average and sum can share the same properties.

danieljbruce committed:
Improve transaction tests

Add sum and average to the transaction tests here to improve test coverage for transactions.

danieljbruce committed:
Change the description in the describe block

danieljbruce committed:
Change return type to average

The return type for this function is wrong. The function returns an Average so we should use Average.

danieljbruce committed:
Make alias optional

Make alias optional since the query still works without providing an alias.

danieljbruce committed:
Fix the transaction tests to fail on rollback

The transaction tests should not pass if the transaction is rolled back like they do currently. We must not catch errors and instead let the test fail.

danieljbruce committed:
Add additional assertions to existing tests

Ensure that the addAggregation function works correctly with an additional assertion check.

danieljbruce committed:
Revert "Add additional assertions to existing tests"

This reverts commit 970c0f412d3c31653c389e9d765817cbb27acb56.

danieljbruce committed:
Add describe block for comparing equivalent query

This test ensures that all the aggregate queries are actually the same no matter how you build them

danieljbruce committed:
Average, sum and count toProto tests

Write tests to effectively document the toProto output of the various aggregate fields

danieljbruce committed:
Add tests for the sum aggregation

Equivalent tests to count are written for sum in system tests and some more tests are written too to meet requirements outlined by team.

vicb commented on src/aggregate.ts:
```diff
@@ -157,4 +196,58 @@ class Count extends AggregateField {
+   *
+   */
+  toProto(): any {
+    const aggregation = Object.assign(
```
Could you just
```js
const aggregation = this.property_ ? {property: {name: this.property_}} : {};
```

danieljbruce replied:
done

vicb commented on src/aggregate.ts:
```diff
@@ -157,4 +196,58 @@ class Count extends AggregateField {
+    return Object.assign(
+      {operator: this.operator},
+      this.alias_ ? {alias: this.alias_} : null,
+      aggregationObject
```
You can also drop the intermediate object and replace this line by
```js
{[this.operator]: aggregation}
```

danieljbruce replied:
done

vicb commented on src/aggregate.ts:
```diff
@@ -157,4 +196,58 @@ class Count extends AggregateField {
+   *
+   * @param {string} property
+   */
+  constructor(property: string) {
```
nit: 
```js
constructor(public property_: string) {
```

and remove the fields + the assignment

danieljbruce replied:
done

danieljbruce committed:
Add a test for sum and snapshot reads

The test for sum and snapshot reads should look at the database before the data is created and run the tests based on the database in that state

danieljbruce committed:
Add two test blocks for special cases

Add a test block for a dataset with overflow values and a dataset with NaN values.

danieljbruce committed:
Export aggregate field from the client

Aggregate field should be exported from the client so that it can be used easily by users.

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-datastore into sum-avg

# Conflicts:
#	src/index.ts

danieljbruce committed:
PR follow-up changes

Some idiomatic changes to improve the state of the code in the PR.

danieljbruce committed:
Adjust the values so that tests pass

Values for sum and average should be different from those of count and these tests provide the right values now.

# Conflicts:
#	system-test/datastore.ts

danieljbruce committed:
Add average aggregations

Average aggregations regarding appearances have been added in tests and correct values have been assigned

danieljbruce committed:
Add snapshot reads for run query and aggregate q

The future refactor must implement the TODOs so that there is less repeated code in the codebase. Also, this commit implements snapshot reads for queries and adds a test for the snapshot reads.

danieljbruce committed:
Remove Google error and entity filter

Remove some unused imports as they do not apply to the code anymore

danieljbruce committed:
Should use null for an aggregation query read time

Snapshot reads read at a time that there is no data so sums and averages should reflect that accordingly.

danieljbruce committed:
Remove tests from a bad cherry pick

Tests for sum that have values corresponding to count are still there in the test cases. They should be removed.

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-datastore into sum-avg

danieljbruce committed:
Linting fix

We don’t care about a loss of precision since the literal value indicated is contained in a test and the loss of precision won’t affect the code.

danieljbruce committed:
Do the test on rating instead of appearances

At this point the datastore is populated with data about ratings so computations should be done on that instead.

danieljbruce committed:
The assertion says the request should have failed

An assertion error should be thrown so that the test doesn’t pass if the request is successful.

danieljbruce committed:
Add a comment about using limits in test

The query with the limit will include all data points with the lowest appearance values. This is likely desired, but also important to document.

danieljbruce commented on system-test/datastore.ts:
```diff
@@ -1130,6 +1369,110 @@ describe('Datastore', () => {
+        .createAggregationQuery(q)
+        .addAggregation(AggregateField.sum('appearances'));
+      const [results] = await datastore.runAggregationQuery(aggregate);
+      assert.deepStrictEqual(results, [{property_1: -18446744073709552000}]);
```
I can confirm that this value and the value for average are the same values received by the grpc call so the values you see here are strictly backend calculations.

danieljbruce committed:
Add rollbacks to transaction tests

The rollbacks for the transaction tests ensure that if a test fails then the data gets reset to where it was before.

danieljbruce commented on src/request.ts:
```diff
@@ -891,13 +891,26 @@ class DatastoreRequest {
+      sharedQueryOpts.readOptions.readTime = {
+        seconds: Math.floor(seconds),
+      };
+    }
```
This code was taken from https://github.com/googleapis/nodejs-datastore/blob/e7c3aaf665ebae5a91fb692bb0af54717e5ea587/src/request.ts#L290-L299 and can be refactored after the merge as per the TODO.

kolea2 replied:
any reason why we can't refactor this now?

danieljbruce replied:
No reason. It is a choice between having two simpler PRs and one PR that includes both changes. Doing so now is fine though.

kolea2 commented on src/request.ts:
```diff
@@ -891,13 +891,26 @@ class DatastoreRequest {
     query: Query,
     options: RunQueryStreamOptions = {}
   ): SharedQueryOptions {
+    // TODO: This needs to be refactored later:
```
is this to be done in this PR or in a future one?

danieljbruce replied:
Same the above comment. We can do it now if preferred.

kolea2 commented on system-test/datastore.ts:
```diff
@@ -704,6 +710,13 @@ describe('Datastore', () => {
       assert.strictEqual(secondEntities!.length, 3);
     });
 
+    it('should query the datastore with snapshot read', async () => {
```
is this related to SUM/AVG changes? If not, let's move it into a separate PR

danieljbruce replied:
ok

kolea2 commented on system-test/datastore.ts:
```diff
@@ -950,6 +963,235 @@ describe('Datastore', () => {
+        const aggregate = datastore
+          .createAggregationQuery(q)
+          .addAggregations([
+            AggregateField.sum('appearances'),
```
might be more interesting to include a sum, average, and a count all in one query

danieljbruce replied:
I could do that in a separate block. This test is inside the `with a sum filter` block.

danieljbruce replied:
Done

danieljbruce commented on system-test/datastore.ts:
```diff
@@ -1130,6 +1372,110 @@ describe('Datastore', () => {
+    after(async () => {
+      await datastore.delete(keys);
+    });
+    it('should run a sum aggregation', async () => {
```
Include `overflow` word here

danieljbruce replied:
Done

danieljbruce committed:
refactor getSharedOptionsOnly

Introducing getSharedOptionsOnly allows us to use that function in two different places to avoid a repeated block of code.

danieljbruce committed:
Remove test related to snapshot reads

This test belongs inside another PR because it is not directly related to sum/avg.

danieljbruce committed:
Add a test for multiple types of aggregates

A test should be included that looks at multiple aggregations in a single query.

danieljbruce committed:
Correct descriptions of two tests on overflow

The tests themselves should include the word overflow so that it is clear they are working with an overflow dataset.

danieljbruce committed:
Merge branch 'main' into sum-avg

kolea2 commented on src/aggregate.ts:
```diff
@@ -121,6 +144,24 @@ abstract class AggregateField {
+   * @returns {Average}
+   */
+  static average(property: string): Sum {
+    return new Average(property);
```
this looks done, good to resolve?

kolea2 commented on src/aggregate.ts:
```diff
@@ -121,14 +144,32 @@ abstract class AggregateField {
+  }
+
   /**
    * Gets a copy of the Count aggregate field.
```
should this be updated to remove "Count" since it applies to all?

danieljbruce replied:
Done

kolea2 commented on src/aggregate.ts:
```diff
@@ -121,14 +144,32 @@ abstract class AggregateField {
    * @returns {AggregateField}
    */
-  alias(alias: string): AggregateField {
+  alias(alias?: string): AggregateField {
```
can we add some additional testing on Count now that this is optional?

danieljbruce replied:
Yes. Added in commit called "Added tests for when an empty alias is provided".

kolea2 commented on src/request.ts:
```diff
@@ -887,9 +867,8 @@ class DatastoreRequest {
-  private getSharedQueryOptions(
-    query: Query,
-    options: RunQueryStreamOptions = {}
+  private getSharedOptionsOnly(
```
what is the difference between this method and `getSharedQueryOptions`? The naming is confusing here.

danieljbruce replied:
`getSharedQueryOptions` also accepts a query and calls `getSharedOptionsOnly`. `getSharedOptionsOnly` is also called for fetches in `createReadStream` where options need to be built without considering a query.

kolea2 replied:
Thanks for the explanation! Maybe `getSharedQueryOptions` and `getSharedRequestOptions`, or `getQueryOptions` and `getRequestOptions` would be a little more readable, wdyt?

danieljbruce replied:
`getQueryOptions` and `getRequestOptions` would work really well. The good thing about these two functions is that they are both private so we can change the names later too if need be.

kolea2 commented on system-test/datastore.ts:
```diff
@@ -676,12 +677,17 @@ describe('Datastore', () => {
     ];
 
     before(async () => {
+      function sleep(ms: number) {
```
why are we adding this? Can you add a comment above explaining?

kolea2 commented on system-test/datastore.ts:
```diff
@@ -1014,6 +1249,22 @@ describe('Datastore', () => {
+        const aggregate = datastore
+          .createAggregationQuery(q)
+          .addAggregations([
+            AggregateField.count(),
```
I would add one more test here using a combination of aliases and non aliases (aka count with an alias, sum without an alias, avg with an alias all in one query)

danieljbruce replied:
Done

danieljbruce replied:
https://github.com/googleapis/nodejs-datastore/pull/1097/commits/bbff19ff13c0eabb19dee64536b2e493d5982cf9

danieljbruce committed:
Add a comment for setting the alias

The comment for setting the alias should not make any mention of count since it is agnostic to the aggregation type.

danieljbruce committed:
Add tests to compare various ways to encode alias

No matter how alias is encoded, the data structures should store the aggregations the same way inside an aggregate field so as not to create any confusion.

danieljbruce committed:
Added tests for when an empty alias is provided

Tests for when an empty alias is provided should check that each aggregate query still works.

danieljbruce committed:
Add a comment clarifying the use of snapshot reads

The sleep function should enable us to test snapshot reads for aggregate queries

danieljbruce committed:
Add two tests to explore mixed aggregations alias

Two tests should be explored that evaluate what happens when multiple aggregations are used and when too many aggregations are used.

danieljbruce committed:
Better names for some internal private functions

Shared options functions could be given a better name so that they make more sense to the code that is using them.

ruyadorno commented on system-test/datastore.ts:
```diff
@@ -676,12 +677,18 @@ describe('Datastore', () => {
     ];
 
     before(async () => {
+      // This function is used for testing snapshot reads by getting a time and then sleeping before adding data.
```
I believe it would be important to rephrase this comment so that it explains **why** we need the sleeping: 
is it simulating process time from datastore? is it simulating latency?

ruyadorno reviewed (commented):
A couple of questions and tweaking a comment but otherwise it's looking good to me 👍

danieljbruce commented on src/index.ts:
```diff
@@ -40,8 +40,9 @@ import * as is from 'is';
 import {entity, Entities, Entity, EntityProto, ValueProto} from './entity';
+import {AggregateField} from './aggregate';
 import Key = entity.Key;
-export {Entity, Key};
```
See design doc comment.

ruyadorno commented on src/index.ts:
```diff
@@ -1818,7 +1819,6 @@ promisifyAll(Datastore, {
     'isDouble',
     'geoPoint',
     'getProjectId',
-    'getSharedQueryOptions',
```
makes sense then, so it sounds like `getSharedQueryOptions` was added here in error since it also lived in `request.ts`

danieljbruce replied:
Yes. I think so.

danieljbruce committed:
Add a comment explaining why the sleep is needed

The code must explain why the sleep function is needed in the test because its purpose should be clear.

danieljbruce committed:
Add getReadTime function and use for sum/avg

A sum/average test uses snapshot reads for an aggregate query. The key here is write code that will guarantee the read time occurs well before all the data is saved in order to ensure the test isn’t flakey.

kolea2 commented on system-test/datastore.ts:
```diff
@@ -689,8 +709,16 @@ describe('Datastore', () => {
-      timeBeforeDataCreation = Date.now();
-      await sleep(1000);
+      // Save for a key so that a read time can be accessed for snapshot reads.
+      const dummyData = Object.assign(Object.assign({}, keysToSave[0]), {
```
rename to "emptyData"

danieljbruce replied:
Done

danieljbruce committed:
Rename variable to emptyData

emptyData is a more accurate variable name for the datastore save data

kolea2 reviewed (approved):
lgtm, please wait for @ruyadorno's approval as well before merging. Thanks for all the hard work on this!

ruyadorno reviewed (commented):
LGTM 👍
